src/invariant.ts

```diff
@@ -0,0 +1,330 @@
+import {
+  AddFeeTier,
+  ChangeFeeReceiver,
+  ChangeProtocolFee,
+  ClaimFee,
+  CreatePool,
+  CreatePosition,
+  Invariant as InvariantFactory,
+  InvariantInstance,
+  RemoveFeeTier,
+  RemovePosition,
+  Swap,
+  TransferPosition,
+  WithdrawProtocolFee
+} from '../artifacts/ts'
+import { FeeTier, Pool, PoolKey, Position, QuoteResult, Tick } from '../artifacts/ts/types'
+import { calculateTick } from './math'
+import { Network } from './network'
+import { getReserveAddress } from './testUtils'
+import {
+  balanceOf,
+  decodeFeeTiers,
+  decodePool,
+  decodePosition,
+  decodeTick,
+  deployCLAMM,
+  deployReserve,
+  MAP_ENTRY_DEPOSIT,
+  waitTxConfirmed
+} from './utils'
+import { Address, DUST_AMOUNT, SignerProvider } from '@alephium/web3'
+
+export class Invariant {
+  instance: InvariantInstance
+  network: Network
+  address: Address
+
+  private constructor(address: Address, network: Network) {
+    this.address = address
+    this.instance = InvariantFactory.at(address)
+    this.network = network
+  }
+
+  static async deploy(
+    signer: SignerProvider,
+    network: Network,
+    protocolFee: bigint = 0n
+  ): Promise<Invariant> {
+    const account = await signer.getSelectedAccount()
+    const clamm = await deployCLAMM(signer)
+    const reserve = await deployReserve(signer)
```

Comment:
`Promise.all` here? Can you deploy both contracts simultaneously, or is there some `nonce` variable that will block this?
```Typescript
    const clamm = await deployCLAMM(signer)
    const reserve = await deployReserve(signer)
```